### PR TITLE
Major refactor.

### PR DESCRIPTION
**Pull request by tomaskulich**
For the sake of my project I significantly rewrote the calendar component. Please inspect the code and consider whether you'd like to accept the changes.

Also, I'm sorry for the monolithic PR, but I make the calendar internal part of my project, so I had to 'cut it out' somehow. 

Finally, the PR is not 100% ready yet; for example the CalendarExample should be adjusted a little - if you consider my PR worth merging, I'll prepare it to the final state then.

List of changes:
- Cleaning & documenting various part of the project
- Significant increase in rendering month view speed (from ~450ms to ~2ms)
- Day component and styles factored out to separate files
- Ability to provide custom (localized) month-names
- Ability to specify, whether week starts on Sunday, Monday, ...
- Using more ES6 features (especially, let and const)
- Eslint

### CODE REVIEW

**Comment by christopherdro** [comment]
@tomaskulich Thank you submitting this PR!

I look forward to a lot of the changes presented. I should have some time next weekend to give this a look.

**Comment by tomaskulich** [comment]
Perfect! Please let me know, If you'll need any explanation from my side.

**Comment by nillo** [comment]
Hi there, i was looking at your code, i hade a question though: 
require('ReactPropTypes') what exactly are u requiring, i thought perhaps u ment
const {
  PropTypes,
} = React; ?

**Comment by tomaskulich** [comment]
uh oh, that looks like a stupid mistake :) It's interesting that the component works at all :) I'll correct it.

**Comment by tomaskulich** [comment]
@christopherdro are you interested in merging the PR?

**Comment by andrejskok** [comment]
@christopherdro I fixed the pull request, should be ok now. Please take a look at it, if you are interested in merging. Thank you

**Comment by AaronO** [comment]
@christopherdro @coderdave Any chance we can get this merged ?

The rendering performance is really really slow on my side without this, and is much improved with this PR. 

I'm happy to rebase this PR on the current master if necessary or submit a new PR that has these performance improvements and few code improvements

**Comment by coderdave** [comment]
Ok, if someone would like to rebase this PR with the current master, I can try to take a look. The only thing I'm not too crazy about is the linter file.

**Comment by tomaskulich** [comment]
We will do the rebasing and also we'll adjust docs and exmaple. We'll do this during the weekend so you can expect this being done on Monday 23.5.

As for the eslint, @coderdave you generally don't like eslint, or you're just not happy with our config? If this is the case, how should the config look like?

**Comment by christopherdro** [comment]
Hey guys. Sorry for being MIA. Just been deep in a project for the last months.

I like the idea of adding a eslintrc lets just simplify by maybe just extended react for now.

@tomaskulich Than you so much sticking through with this!

**Comment by christopherdro** [comment]
This component was made before React Native Android was every announced, hence the file names using .ios. I think we can get rid of all that and just Calender.js and Day.js.

Thoughts?

**Comment by tomaskulich** [comment]
@christopherdro: 

> simplify by maybe just extended react for now.
> Do you have in mind some standardized config, we could extend? Which one?

ad .ios.js -> .js : I'll be more than happy to do that. We'll cover also that during the weekend.

**Comment by christopherdro** [comment]
👍 Yes. Calendar.js, Day.js

**Comment by christopherdro** [comment]
I've been thinking lately about adding some native features
like pulling/adding to native calendars. I think that could be useful.

**Comment by coderdave** [comment]
If we're going to use eslintrc, I like the one that Abramov has come up with:
https://github.com/gaearon/library-boilerplate/blob/master/.eslintrc

Maybe that can be the starting point for now?

**Comment by christopherdro** [comment]
Sounds great!

**Comment by tomaskulich** [comment]
ok, so Dan Abramov's (well, it basically inherits everything from airbnb) .eslintrc it is.

**Comment by andrejskok** [comment]
Hi guys, i edited the eslint and fixed errors and warning it caused. I got rid of some deprecated methods and set minimum RN version to 0.13.0 which is quite reasonable these days. I updated the README.md and CalendarExample, please check those too and let me know if there is something more to fix. My CalendarExample is importing calendar sources from ./src directory in the project, not installing through npm, bacause of the slightly changed API, and i needed to test it. I am aware that this will have to be fixed.

**Comment by coderdave** [comment]
Github is saying that there's conflicts that must be resolved? Can you merge in master and fix them?

Also, can we change `const React = require('react-native');` to `const ReactNative = require('react-native');` (or something similar) so the destructuring assignment is pointing to react native and not react so as to not be confusing?

Or, we can be cool and start using import:
`import { PropTypes } from 'react';`
`import { Dimensions, ScrollView, Text, TouchableOpacity, View } from 'react-native';`

**Comment by christopherdro** [comment]
I would suggest updating the import since that is now a requirement with the current version of react native.

**Comment by coderdave** [comment]
Some of the "internal" function names are prefixed with underscore and some aren't. I suggest we change them all to use underscores or not use underscores at all. (Or maybe I'm missing something here)

Also, what happened to `onSwipeNext` and `onSwipePrev`? They're not being called.

Another thing I just noticed - `onTouchPrev` is being called on BOTH `_onPrev()` and `_onNext()`

File: calendar.js [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+  _onPrev() {
+    const newMoment = moment(this.state.currentMonthMoment).add(-1, 'month');
+    this.setState({ currentMonthMoment: newMoment });
+    this.props.onTouchPrev && this.props.onTouchPrev(newMoment); // eslint-disable-line no-unused-expressions, max-len
+  }
+
+  _onNext() {
+    const newMoment = moment(this.state.currentMonthMoment).add(1, 'month');
+    this.setState({ currentMonthMoment: newMoment });
+    this.props.onTouchPrev && this.props.onTouchPrev(newMoment); // eslint-disable-line no-unused-expressions, max-len
```
Comment (christopherdro) [inline calendar.js]: @coderdave Good point this should call the onTouchNext prop

File: calendar.js [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+  _selectDate(date) {
+    this.setState({
+      selectedMoment: date,
+    });
+    this.props.onDateSelect && this.props.onDateSelect(date.format()); // eslint-disable-line no-unused-expressions, max-len
+  }
+
+  _onPrev() {
+    const newMoment = moment(this.state.currentMonthMoment).add(-1, 'month');
```
Comment (christopherdro) [inline calendar.js]: Why use `add -1` instead of subtracting 1 like the original implementation

**Comment by andrejskok** [comment]
Updated and rebased.

**Pull request merged**